_posts/2020-07-29-parallelized-sampling/parallelized-sampling.Rmd

```diff
@@ -0,0 +1,290 @@
+
+```{r}
+dplyr::sample_frac(mtcars, 0.25, weight = gear, replace = FALSE)
+```
+
+```
+##                    mpg cyl  disp  hp drat    wt  qsec vs am gear carb
+## Merc 280C         17.8   6 167.6 123 3.92 3.440 18.90  1  0    4    4
+## Chrysler Imperial 14.7   8 440.0 230 3.23 5.345 17.42  0  0    3    4
+## Fiat X1-9         27.3   4  79.0  66 4.08 1.935 18.90  1  1    4    1
+## Hornet Sportabout 18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2
+## Valiant           18.1   6 225.0 105 2.76 3.460 20.22  1  0    3    1
+## Porsche 914-2     26.0   4 120.3  91 4.43 2.140 16.70  0  1    5    2
+## Maserati Bora     15.0   8 301.0 335 3.54 3.570 14.60  0  1    5    8
+## Ferrari Dino      19.7   6 145.0 175 3.62 2.770 15.50  0  1    5    6
+```
+
+will randomly select one-fourth of all rows from a R dataframe named "mtcars" without replacement, using `mtcars$gear` as weights. We were unable to find any function implementing the weighted versions of `dplyr::sample_frac` among [Spark SQL built-in functions](https://spark.apache.org/docs/3.0.0/api/sql/index.html) in Spark 3.0 or in earlier versions, which means a future version of `sparklyr` will need to run its own weighted sampling algorithm to support such use cases.
+
+## What exactly is `SWoR` {#swor}
+
+The purpose of this section is to mathematically describe the probability distribution generated by `SWoR` in terms of $w_1, \dotsc, w_N$, so that readers can clearly see that the exponential-variate based algorithm presented in a subsequent section in fact samples from precisely the same probability distribution. Readers already having a crystal-clear mental picture of what `SWoR` entails should probably skip most of this section. The key take-away here is given $N$ rows $r_1, \dotsc, r_N$ and their weights $w_1, \dotsc, w_N$ and a desired sample size $n$, the probability of `SWoR` selecting $(r_1, \dotsc, r_n)$ is $\prod\limits_{j = 1}^{n} \left( {w_j} \middle/ {\sum\limits_{k = j}^{N}{w_k}} \right)$.
+
+`SWOR` is conceptually equivalent to a $n$-step process of selecting 1 out of $(n - j + 1)$ remaining rows in the $j$-th step for $j \in \{1, \dotsc, n\}$, with each remaining row's likelihood of getting selected being linearly proportional to its weight in any of the steps, i.e.,
+
+```
+samples := {}
+population := {r[1], ..., r[N]}
+
+for j = 1 to n
+  select r[x] from population with probability
+    (w[x] / TotalWeight(population))
+  samples := samples + {r[x]}
+  population := population - {r[x]}
+```
+
+Notice the outcome of a `SWoR` process is in fact order-significant, which is why in this post it will always be represented as an ordered tuple of elements.
+
+Intuitively, `SWoR` is analogous to throwing darts at a bunch of tiles. For example, let's say the size of our sample space is 5:
```

Comment:
Not gonna lie, I'm a little distracted by the vision of a dart being thrown at a tile. since it would just smack it and fall to the floor, but I don't think you actually need to change anything here. 🎯